Feat/rest api impl

rivon0507 opened the pull request:
Closes #4 

This pull request introduces a comprehensive authentication system, along with supporting changes to enhance security, user management, and exception handling. The most significant updates include the implementation of JWT-based authentication, the addition of services for user registration and login, and modifications to entity IDs for consistency. Below are the key changes grouped by theme.

### Authentication and Security
* Added `SecurityConfig` to configure Spring Security, enabling JWT-based stateless authentication, and defining public and protected endpoints (`/login`, `/register`). 
* Implemented `JwtAuthenticationFilter` to validate JWT tokens and set up the security context for authenticated requests. 
* Added `JwtService` to handle JWT generation, validation, and extraction of claims, using a secret key and expiration settings. 

### User Management
* Created `AuthService` to handle user registration and login, including password encoding, email uniqueness checks, and token generation. 
* Implemented `CustomUserDetailsService` to load user details by email for authentication purposes. 
* Added `UserRepository` to support user-related database operations, such as email existence checks and retrieval by email. 

### Entity and DTO Enhancements
* Changed entity IDs (`Conversation`, `Member`, `Message`, and `User`) from `UUID` to `Long` for consistency and easier integration with other systems.
* Added a static factory method `from` in `UserDTO` to convert a `User` entity into a `UserDTO`. 

### Exception Handling
* Introduced `EmailTakenException` for handling cases where a duplicate email is used during registration. 
* Updated `GlobalExceptionHandler` to handle `EmailTakenException` and `AuthenticationException`, returning appropriate HTTP status codes and error details. 

### API Documentation
* Updated the base URL in the REST API documentation to include `/api/v1`, reflecting the new API versioning structure.

rivon0507 committed:
Auth implementation

Ryan2486 commented on src/main/java/com/_2rkdev/dischard/dto/common/UserDTO.java:
For better readablility, separe the mapper and DTO

Ryan2486 reviewed (dismissed):
Need some refactoring but it s okey

Ryan2486 commented on src/main/java/com/_2rkdev/dischard/service/AuthService.java:
```diff
@@ -0,0 +1,53 @@
+    public LoginResponseDTO login(@NotNull LoginRequestDTO request) {
+        authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(request.email(), request.password()));
+        String token = jwtService.generateToken(userDetailsService.loadUserByUsername(request.email()));
+        return new LoginResponseDTO(token, UserDTO.from(userRepository.findByEmail(request.email()).orElseThrow()));
```
that s right

AmigosKazz reviewed (dismissed):
Gg 👍 !

rivon0507 committed:
Apply suggestions from code review

- throw a BadCredentialsException when User not found in AuthService
- user UserMapper to map User entity to UserDTO
- remove redundant many-to-many relationship between Member and Conversation => one-to-many
- single table inheritance for Message
- replace @Data annotation of @Entities with overridden equals and hashCode

rivon0507 committed:
Fix build error with duplicate constructor definition

rivon0507 committed:
Add Integration tests for auth controller

rivon0507 committed:
Add edge case tests to justify not annotating LoginRequestDTO with @Valid :P
